pkg/common/cliprinter/structjson/structjson.go

```diff
@@ -0,0 +1,30 @@
+package structjson
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+func Print(msgs []interface{}, stdout, _ io.Writer) {
+	var jb []byte
+	var err error
+
+	if len(msgs) == 0 {
+		return
+	}
+
+	// TODO: Should we do this?
+	// If we're called from a `show` command, it makes sense
+	// but if a list operation has only one result, it does not
+	if len(msgs) == 1 {
+		jb, err = json.Marshal(msgs[0])
+	} else {
+		jb, err = json.Marshal(msgs)
+	}
+	if err != nil {
```

Comment:
Seems like errors should go to stderr...

Reply:
yea ..

I've been going back and forth on how to handle errors here in general. I think your observation about signaling the Print() caller so we can exit non-zero is a good one... I think it is also important to print errors with the requested formatting. We can print to stderr with the requested formatting when we encounter errors like this, or we can return the formatted error to the caller and hope they don't mangle it ...

Maybe we should always print to stderr with the requested formatting, and return only a bool to the Print() caller so that they don't have to handle printing anything further.

Reply:
I think that sounds ok. The important bit to me is errors go to stderr and an exit with non-zero status. A think formatting of the error is less important since typically one is only piping stdout to things but aligning on the requested format seems fine if it's easy to do.

Reply:
> The important bit to me is errors go to stderr and an exit with non-zero status.

Non-zero certainly.

When formatting with something other than pretty print though, where the output is being consumed by a non-human .. do you want the error notification to go to the consumer, or the executor? E.g. does Cron get to know about the error, or does script that processes the job output get to know about it? Or both?

> A think formatting of the error is less important since typically one is only piping stdout to things but aligning on the requested format seems fine if it's easy to do.

Believe it or not, I have been bit by this before!

Reply:
Fair enough!